File: platform/cli/src/commands/clone.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as path from 'node:path';
import * as fs from 'node:fs';
import * as git from '../git';
import { ensureDevpodDir } from '../workspace';
import { errorMessage, nextStepHint } from '../format';

export function registerCloneCommand(program: Command): void {
  program
    .command('clone')
    .description('Clone a repository and set up devpod')
    .argument('<repo>', 'Repository URL or owner/name')
    .action(async (repo: string) => {
      try {
        // Resolve repo URL
        let url = repo;
        if (!repo.includes('://') && !repo.startsWith('git@')) {
          url = `https://github.com/${repo}.git`;
        }

        // Derive directory name from URL
        const repoName = url
          .replace(/\.git$/, '')
          .split('/')
          .pop() || 'repo';

        console.log(`Cloning ${chalk.cyan(repo)}...`);

        git.clone(url, repoName);

        // Initialize .devpod workspace inside the cloned repo
        const fullPath = path.resolve(process.cwd(), repoName);
        ensureDevpodDir(fullPath);

        // Add .devpod/ to .git/info/exclude
        const excludePath = path.join(fullPath, '.git', 'info', 'exclude');
        try {
          const excludeDir = path.dirname(excludePath);
          if (!fs.existsSync(excludeDir)) {
            fs.mkdirSync(excludeDir, { recursive: true });
          }
          const existing = fs.existsSync(excludePath)
            ? fs.readFileSync(excludePath, 'utf-8')
            : '';
          if (!existing.includes('.devpod/')) {
            fs.writeFileSync(excludePath, existing.trimEnd() + '\n.devpod/\n');
          }
        } catch {
          // Non-fatal: exclude file couldn't be written
        }

        console.log(`${chalk.green('\u2713')} Cloned to ${chalk.bold(repoName)}`);
        console.log(chalk.dim(`  cd ${repoName}`));
        console.log(nextStepHint('clone'));
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}

File: platform/cli-go/internal/commands/clone.go
package commands

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/git"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

func newCloneCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "clone <repo>",
		Short: "Clone a repository and set up devpod",
		Long: `Clone a repository and initialize the devpod workspace.

Each feature you create will have two branches:
  feature/<slug>           Your working branch (clean commits)
  feature/<slug>--versions  Append-only history of every save`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			repo := args[0]

			// Resolve repo URL
			url := repo
			if !strings.Contains(repo, "://") && !strings.HasPrefix(repo, "git@") {
				url = "https://github.com/" + repo + ".git"
			}

			// Derive directory name from URL
			repoName := url
			repoName = strings.TrimSuffix(repoName, ".git")
			parts := strings.Split(repoName, "/")
			repoName = parts[len(parts)-1]
			if repoName == "" {
				repoName = "repo"
			}

			fmt.Printf("Cloning %s...\n", repo)

			if err := git.Clone(url, repoName); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			// Initialize .devpod workspace inside the cloned repo
			cwd, _ := os.Getwd()
			fullPath := filepath.Join(cwd, repoName)
			if err := workspace.EnsureDevpodDir(fullPath); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
			}

			// Add .devpod/ to .git/info/exclude
			excludePath := filepath.Join(fullPath, ".git", "info", "exclude")
			excludeDir := filepath.Dir(excludePath)
			_ = os.MkdirAll(excludeDir, 0o755)
			existing, _ := os.ReadFile(excludePath)
			if !strings.Contains(string(existing), ".devpod/") {
				content := strings.TrimRight(string(existing), "\n") + "\n.devpod/\n"
				_ = os.WriteFile(excludePath, []byte(content), 0o644)
			}

			fmt.Printf("%s Cloned to %s\n", format.SuccessMsg("\u2713"), repoName)
			fmt.Printf("  cd %s\n", repoName)
			fmt.Println(format.NextStepHint("clone"))
			return nil
		},
	}
}
